The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18650848065/job/53168314542) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/53168314542) [(plain)](https://github.com/rust-lang/rust/commit/316479ae3bc55c6c1ecc0f67a54bdd9eeee0c085/checks/53168314542/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
fmt: checked 6501 files
tidy check
tidy [rustdoc_json (src)]: `rustdoc-json-types` modified, checking format version
tidy: Skipping binary file check, read-only filesystem
tidy [style (tests)]: /checkout/tests/ui/traits/next-solver/unsize-goal-mismatch.rs:7: line longer than 100 chars
tidy [style (tests)]: FAIL
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.2)
linting python files
---
    |
759 |                     // writen in source somewhere (otherwise we will never have lowered
    |                        ^^^^^^
    |
rerun tidy with `--extra-checks=spellcheck --bless` to fix typos
tidy [extra_checks]: checks with external tool 'typos' failed
tidy [extra_checks]: FAIL
tidy: The following checks failed: extra_checks, style (tests)
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy /checkout /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo /checkout/obj/build 4 /node/bin/npm --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1549:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1280:29

Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:02:34
  local time: Mon Oct 20 11:47:08 UTC 2025
  network time: Mon, 20 Oct 2025 11:47:08 GMT
##[error]Process completed with exit code 1.
```

</details>

--- a/compiler/rustc_hir_typeck/src/coercion.rs
+++ b/compiler/rustc_hir_typeck/src/coercion.rs
@@ -35,13 +35,13 @@
 //! // and are then unable to coerce `&7i32` to `&mut i32`.
 //! ```
 
-use std::ops::Deref;
+use std::ops::{ControlFlow, Deref};
 
 use rustc_errors::codes::*;
 use rustc_errors::{Applicability, Diag, struct_span_code_err};
-use rustc_hir as hir;
 use rustc_hir::attrs::InlineAttr;
 use rustc_hir::def_id::{DefId, LocalDefId};
+use rustc_hir::{self as hir, LangItem};
 use rustc_hir_analysis::hir_ty_lowering::HirTyLowerer;
 use rustc_infer::infer::relate::RelateResult;
 use rustc_infer::infer::{DefineOpaqueTypes, InferOk, InferResult, RegionVariableOrigin};
@@ -56,6 +56,8 @@ use rustc_middle::ty::error::TypeError;
 use rustc_middle::ty::{self, GenericArgsRef, Ty, TyCtxt, TypeVisitableExt};
 use rustc_span::{BytePos, DUMMY_SP, DesugaringKind, Span};
 use rustc_trait_selection::infer::InferCtxtExt as _;
+use rustc_trait_selection::solve::inspect::{self, InferCtxtProofTreeExt, ProofTreeVisitor};
+use rustc_trait_selection::solve::{Certainty, Goal, NoSolution};
 use rustc_trait_selection::traits::query::evaluate_obligation::InferCtxtExt;
 use rustc_trait_selection::traits::{
     self, ImplSource, NormalizeExt, ObligationCause, ObligationCauseCode, ObligationCtxt,
@@ -639,131 +641,140 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
             Adjust::Pointer(PointerCoercion::Unsize),
         )?;
 
-        let mut selcx = traits::SelectionContext::new(self);
-
         // Create an obligation for `Source: CoerceUnsized<Target>`.
         let cause = self.cause(self.cause.span, ObligationCauseCode::Coercion { source, target });
+        let pred = ty::TraitRef::new(self.tcx, coerce_unsized_did, [coerce_source, coerce_target]);
+        let obligation = Obligation::new(self.tcx, cause, self.fcx.param_env, pred);
 
-        // Use a FIFO queue for this custom fulfillment procedure.
-        //
-        // A Vec (or SmallVec) is not a natural choice for a queue. However,
-        // this code path is hot, and this queue usually has a max length of 1
-        // and almost never more than 3. By using a SmallVec we avoid an
-        // allocation, at the (very small) cost of (occasionally) having to
-        // shift subsequent elements down when removing the front element.
-        let mut queue: SmallVec<[PredicateObligation<'tcx>; 4]> = smallvec![Obligation::new(
-            self.tcx,
-            cause,
-            self.fcx.param_env,
-            ty::TraitRef::new(self.tcx, coerce_unsized_did, [coerce_source, coerce_target])
-        )];
+        if self.next_trait_solver() {
+            coercion.obligations.push(obligation);
 
-        // Keep resolving `CoerceUnsized` and `Unsize` predicates to avoid
-        // emitting a coercion in cases like `Foo<$1>` -> `Foo<$2>`, where
-        // inference might unify those two inner type variables later.
-        let traits = [coerce_unsized_did, unsize_did];
-        while !queue.is_empty() {
-            let obligation = queue.remove(0);
-            let trait_pred = match obligation.predicate.kind().no_bound_vars() {
-                Some(ty::PredicateKind::Clause(ty::ClauseKind::Trait(trait_pred)))
-                    if traits.contains(&trait_pred.def_id()) =>
-                {
-                    self.resolve_vars_if_possible(trait_pred)
-                }
-                // Eagerly process alias-relate obligations in new trait solver,
-                // since these can be emitted in the process of solving trait goals,
-                // but we need to constrain vars before processing goals mentioning
-                // them.
-                Some(ty::PredicateKind::AliasRelate(..)) => {
-                    let ocx = ObligationCtxt::new(self);
-                    ocx.register_obligation(obligation);
-                    if !ocx.try_evaluate_obligations().is_empty() {
-                        return Err(TypeError::Mismatch);
-                    }
-                    coercion.obligations.extend(ocx.into_pending_obligations());
-                    continue;
-                }
-                _ => {
-                    coercion.obligations.push(obligation);
-                    continue;
-                }
-            };
-            debug!("coerce_unsized resolve step: {:?}", trait_pred);
-            match selcx.select(&obligation.with(selcx.tcx(), trait_pred)) {
-                // Uncertain or unimplemented.
-                Ok(None) => {
-                    if trait_pred.def_id() == unsize_did {
-                        let self_ty = trait_pred.self_ty();
-                        let unsize_ty = trait_pred.trait_ref.args[1].expect_ty();
-                        debug!("coerce_unsized: ambiguous unsize case for {:?}", trait_pred);
-                        match (self_ty.kind(), unsize_ty.kind()) {
-                            (&ty::Infer(ty::TyVar(v)), ty::Dynamic(..))
-                                if self.type_var_is_sized(v) =>
-                            {
-                                debug!("coerce_unsized: have sized infer {:?}", v);
-                                coercion.obligations.push(obligation);
-                                // `$0: Unsize<dyn Trait>` where we know that `$0: Sized`, try going
-                                // for unsizing.
-                            }
-                            _ => {
-                                // Some other case for `$0: Unsize<Something>`. Note that we
-                                // hit this case even if `Something` is a sized type, so just
-                                // don't do the coercion.
-                                debug!("coerce_unsized: ambiguous unsize");
-                                return Err(TypeError::Mismatch);
-                            }
-                        }
-                    } else {
-                        debug!("coerce_unsized: early return - ambiguous");
-                        return Err(TypeError::Mismatch);
-                    }
-                }
-                Err(SelectionError::Unimplemented) => {
-                    debug!("coerce_unsized: early return - can't prove obligation");
-                    return Err(TypeError::Mismatch);
-                }
+            if self
+                .infcx
+                .visit_proof_tree(
+                    Goal::new(self.tcx, self.param_env, pred),
+                    &mut CoerceVisitor { fcx: self.fcx, span: self.cause.span },
+                )
+                .is_break()
+            {
+                return Err(TypeError::Mismatch);
+            }
+        } else {
+            let mut selcx = traits::SelectionContext::new(self);
+            // Use a FIFO queue for this custom fulfillment procedure.
+            //
+            // A Vec (or SmallVec) is not a natural choice for a queue. However,
+            // this code path is hot, and this queue usually has a max length of 1
+            // and almost never more than 3. By using a SmallVec we avoid an
+            // allocation, at the (very small) cost of (occasionally) having to
+            // shift subsequent elements down when removing the front element.
+            let mut queue: SmallVec<[PredicateObligation<'tcx>; 4]> = smallvec![obligation];
 
-                Err(SelectionError::TraitDynIncompatible(_)) => {
-                    // Dyn compatibility errors in coercion will *always* be due to the
-                    // fact that the RHS of the coercion is a non-dyn compatible `dyn Trait`
-                    // written in source somewhere (otherwise we will never have lowered
-                    // the dyn trait from HIR to middle).
-                    //
-                    // There's no reason to emit yet another dyn compatibility error,
-                    // especially since the span will differ slightly and thus not be
-                    // deduplicated at all!
-                    self.fcx.set_tainted_by_errors(
-                        self.fcx
-                            .dcx()
-                            .span_delayed_bug(self.cause.span, "dyn compatibility during coercion"),
-                    );
-                }
-                Err(err) => {
-                    let guar = self.err_ctxt().report_selection_error(
-                        obligation.clone(),
-                        &obligation,
-                        &err,
-                    );
-                    self.fcx.set_tainted_by_errors(guar);
-                    // Treat this like an obligation and follow through
-                    // with the unsizing - the lack of a coercion should
-                    // be silent, as it causes a type mismatch later.
-                }
-
-                Ok(Some(ImplSource::UserDefined(impl_source))) => {
-                    queue.extend(impl_source.nested);
-                    // Certain incoherent `CoerceUnsized` implementations may cause ICEs,
-                    // so check the impl's validity. Taint the body so that we don't try
-                    // to evaluate these invalid coercions in CTFE. We only need to do this
-                    // for local impls, since upstream impls should be valid.
-                    if impl_source.impl_def_id.is_local()
-                        && let Err(guar) =
-                            self.tcx.ensure_ok().coerce_unsized_info(impl_source.impl_def_id)
+            // Keep resolving `CoerceUnsized` and `Unsize` predicates to avoid
+            // emitting a coercion in cases like `Foo<$1>` -> `Foo<$2>`, where
+            // inference might unify those two inner type variables later.
+            let traits = [coerce_unsized_did, unsize_did];
+            while !queue.is_empty() {
+                let obligation = queue.remove(0);
+                let trait_pred = match obligation.predicate.kind().no_bound_vars() {
+                    Some(ty::PredicateKind::Clause(ty::ClauseKind::Trait(trait_pred)))
+                        if traits.contains(&trait_pred.def_id()) =>
                     {
-                        self.fcx.set_tainted_by_errors(guar);
+                        self.resolve_vars_if_possible(trait_pred)
                     }
+                    // Eagerly process alias-relate obligations in new trait solver,
+                    // since these can be emitted in the process of solving trait goals,
+                    // but we need to constrain vars before processing goals mentioning
+                    // them.
+                    Some(ty::PredicateKind::AliasRelate(..)) => {
+                        let ocx = ObligationCtxt::new(self);
+                        ocx.register_obligation(obligation);
+                        if !ocx.try_evaluate_obligations().is_empty() {
+                            return Err(TypeError::Mismatch);
+                        }
+                        coercion.obligations.extend(ocx.into_pending_obligations());
+                        continue;
+                    }
+                    _ => {
+                        coercion.obligations.push(obligation);
+                        continue;
+                    }
+                };
+                debug!("coerce_unsized resolve step: {:?}", trait_pred);
+                match selcx.select(&obligation.with(selcx.tcx(), trait_pred)) {
+                    // Uncertain or unimplemented.
+                    Ok(None) => {
+                        if trait_pred.def_id() == unsize_did {
+                            let self_ty = trait_pred.self_ty();
+                            let unsize_ty = trait_pred.trait_ref.args[1].expect_ty();
+                            debug!("coerce_unsized: ambiguous unsize case for {:?}", trait_pred);
+                            match (self_ty.kind(), unsize_ty.kind()) {
+                                (&ty::Infer(ty::TyVar(v)), ty::Dynamic(..))
+                                    if self.type_var_is_sized(v) =>
+                                {
+                                    debug!("coerce_unsized: have sized infer {:?}", v);
+                                    coercion.obligations.push(obligation);
+                                    // `$0: Unsize<dyn Trait>` where we know that `$0: Sized`, try going
+                                    // for unsizing.
+                                }
+                                _ => {
+                                    // Some other case for `$0: Unsize<Something>`. Note that we
+                                    // hit this case even if `Something` is a sized type, so just
+                                    // don't do the coercion.
+                                    debug!("coerce_unsized: ambiguous unsize");
+                                    return Err(TypeError::Mismatch);
+                                }
+                            }
+                        } else {
+                            debug!("coerce_unsized: early return - ambiguous");
+                            return Err(TypeError::Mismatch);
+                        }
+                    }
+                    Err(SelectionError::Unimplemented) => {
+                        debug!("coerce_unsized: early return - can't prove obligation");
+                        return Err(TypeError::Mismatch);
+                    }
+
+                    Err(SelectionError::TraitDynIncompatible(_)) => {
+                        // Dyn compatibility errors in coercion will *always* be due to the
+                        // fact that the RHS of the coercion is a non-dyn compatible `dyn Trait`
+                        // writen in source somewhere (otherwise we will never have lowered
+                        // the dyn trait from HIR to middle).
+                        //
+                        // There's no reason to emit yet another dyn compatibility error,
+                        // especially since the span will differ slightly and thus not be
+                        // deduplicated at all!
+                        self.fcx.set_tainted_by_errors(self.fcx.dcx().span_delayed_bug(
+                            self.cause.span,
+                            "dyn compatibility during coercion",
+                        ));
+                    }
+                    Err(err) => {
+                        let guar = self.err_ctxt().report_selection_error(
+                            obligation.clone(),
+                            &obligation,
+                            &err,
+                        );
+                        self.fcx.set_tainted_by_errors(guar);
+                        // Treat this like an obligation and follow through
+                        // with the unsizing - the lack of a coercion should
+                        // be silent, as it causes a type mismatch later.
+                    }
+                    Ok(Some(ImplSource::UserDefined(impl_source))) => {
+                        queue.extend(impl_source.nested);
+                        // Certain incoherent `CoerceUnsized` implementations may cause ICEs,
+                        // so check the impl's validity. Taint the body so that we don't try
+                        // to evaluate these invalid coercions in CTFE. We only need to do this
+                        // for local impls, since upstream impls should be valid.
+                        if impl_source.impl_def_id.is_local()
+                            && let Err(guar) =
+                                self.tcx.ensure_ok().coerce_unsized_info(impl_source.impl_def_id)
+                        {
+                            self.fcx.set_tainted_by_errors(guar);
+                        }
+                    }
+                    Ok(Some(impl_source)) => queue.extend(impl_source.nested_obligations()),
                 }
-                Ok(Some(impl_source)) => queue.extend(impl_source.nested_obligations()),
             }
         }
 
@@ -2003,5 +2014,54 @@ impl AsCoercionSite for ! {
 impl AsCoercionSite for hir::Arm<'_> {
     fn as_coercion_site(&self) -> &hir::Expr<'_> {
         self.body
+    }
+}
+
+/// Recursively visit goals to decide whether an unsizing is possible.
+/// `Break`s when it isn't, and an error should be raised.
+/// `Continue`s when an unsizing ok based on an implementation of the `Unsize` trait / lang item.
+struct CoerceVisitor<'a, 'tcx> {
+    fcx: &'a FnCtxt<'a, 'tcx>,
+    span: Span,
+}
+
+impl<'tcx> ProofTreeVisitor<'tcx> for CoerceVisitor<'_, 'tcx> {
+    type Result = ControlFlow<()>;
+
+    fn span(&self) -> Span {
+        self.span
+    }
+
+    fn visit_goal(&mut self, goal: &inspect::InspectGoal<'_, 'tcx>) -> Self::Result {
+        let Some(pred) = goal.goal().predicate.as_trait_clause() else {
+            return ControlFlow::Continue(());
+        };
+
+        if !self.fcx.tcx.is_lang_item(pred.def_id(), LangItem::Unsize)
+            && !self.fcx.tcx.is_lang_item(pred.def_id(), LangItem::CoerceUnsized)
+        {
+            return ControlFlow::Continue(());
+        }
+
+        match goal.result() {
+            Ok(Certainty::Yes) => ControlFlow::Continue(()),
+            Err(NoSolution) => ControlFlow::Break(()),
+            Ok(Certainty::Maybe { .. }) => {
+                // FIXME: structurally normalize?
+                if self.fcx.tcx.is_lang_item(pred.def_id(), LangItem::Unsize)
+                    && let ty::Dynamic(..) = pred.skip_binder().trait_ref.args.type_at(1).kind()
+                    && let ty::Infer(ty::TyVar(vid)) = *pred.self_ty().skip_binder().kind()
+                    && self.fcx.type_var_is_sized(vid)
+                {
+                    ControlFlow::Continue(())
+                } else if let Some(cand) = goal.unique_applicable_candidate()
+                    && cand.shallow_certainty() == Certainty::Yes
+                {
+                    cand.visit_nested_no_probe(self)
+                } else {
+                    ControlFlow::Break(())
+                }
+            }
+        }
     }
 }

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/inspect_obligations.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/inspect_obligations.rs
@@ -6,7 +6,7 @@ use rustc_middle::ty::{self, Ty, TypeVisitableExt};
 use rustc_span::Span;
 use rustc_trait_selection::solve::Certainty;
 use rustc_trait_selection::solve::inspect::{
-    InspectConfig, InspectGoal, InferCtxtProofTreeExt, ProofTreeVisitor,
+    InferCtxtProofTreeExt, InspectConfig, InspectGoal, ProofTreeVisitor,
 };
 use tracing::{debug, instrument, trace};
 

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/inspect_obligations.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/inspect_obligations.rs
@@ -6,7 +6,7 @@ use rustc_middle::ty::{self, Ty, TypeVisitableExt};
 use rustc_span::Span;
 use rustc_trait_selection::solve::Certainty;
 use rustc_trait_selection::solve::inspect::{
-    InspectConfig, InspectGoal, ProofTreeInferCtxtExt, ProofTreeVisitor,
+    InspectConfig, InspectGoal, InferCtxtProofTreeExt, ProofTreeVisitor,
 };
 use tracing::{debug, instrument, trace};
 

--- a/compiler/rustc_trait_selection/src/solve/fulfill.rs
+++ b/compiler/rustc_trait_selection/src/solve/fulfill.rs
@@ -24,7 +24,7 @@ use tracing::instrument;
 use self::derive_errors::*;
 use super::Certainty;
 use super::delegate::SolverDelegate;
-use super::inspect::{self, ProofTreeInferCtxtExt};
+use super::inspect::{self, InferCtxtProofTreeExt};
 use crate::traits::{FulfillmentError, ScrubbedTraitError};
 
 mod derive_errors;

--- a/compiler/rustc_trait_selection/src/solve/fulfill/derive_errors.rs
+++ b/compiler/rustc_trait_selection/src/solve/fulfill/derive_errors.rs
@@ -15,7 +15,7 @@ use rustc_next_trait_solver::solve::{GoalEvaluation, SolverDelegateEvalExt as _}
 use tracing::{instrument, trace};
 
 use crate::solve::delegate::SolverDelegate;
-use crate::solve::inspect::{self, ProofTreeInferCtxtExt, ProofTreeVisitor};
+use crate::solve::inspect::{self, InferCtxtProofTreeExt, ProofTreeVisitor};
 use crate::solve::{Certainty, deeply_normalize_for_diagnostics};
 use crate::traits::{FulfillmentError, FulfillmentErrorCode, wf};
 

--- a/compiler/rustc_trait_selection/src/solve/inspect/analyse.rs
+++ b/compiler/rustc_trait_selection/src/solve/inspect/analyse.rs
@@ -463,7 +463,7 @@ pub trait ProofTreeVisitor<'tcx> {
     fn visit_goal(&mut self, goal: &InspectGoal<'_, 'tcx>) -> Self::Result;
 }
 
-#[extension(pub trait ProofTreeInferCtxtExt<'tcx>)]
+#[extension(pub trait InferCtxtProofTreeExt<'tcx>)]
 impl<'tcx> InferCtxt<'tcx> {
     fn visit_proof_tree<V: ProofTreeVisitor<'tcx>>(
         &self,

--- a/compiler/rustc_trait_selection/src/solve/select.rs
+++ b/compiler/rustc_trait_selection/src/solve/select.rs
@@ -12,7 +12,7 @@ use rustc_middle::{bug, span_bug};
 use rustc_span::Span;
 use thin_vec::thin_vec;
 
-use crate::solve::inspect::{self, ProofTreeInferCtxtExt};
+use crate::solve::inspect::{self, InferCtxtProofTreeExt};
 
 #[extension(pub trait InferCtxtSelectExt<'tcx>)]
 impl<'tcx> InferCtxt<'tcx> {

--- a/compiler/rustc_trait_selection/src/traits/coherence.rs
+++ b/compiler/rustc_trait_selection/src/traits/coherence.rs
@@ -29,7 +29,7 @@ use tracing::{debug, instrument, warn};
 use super::ObligationCtxt;
 use crate::error_reporting::traits::suggest_new_overflow_limit;
 use crate::infer::InferOk;
-use crate::solve::inspect::{InspectGoal, InferCtxtProofTreeExt, ProofTreeVisitor};
+use crate::solve::inspect::{InferCtxtProofTreeExt, InspectGoal, ProofTreeVisitor};
 use crate::solve::{SolverDelegate, deeply_normalize_for_diagnostics, inspect};
 use crate::traits::query::evaluate_obligation::InferCtxtExt;
 use crate::traits::select::IntercrateAmbiguityCause;

--- a/compiler/rustc_trait_selection/src/traits/coherence.rs
+++ b/compiler/rustc_trait_selection/src/traits/coherence.rs
@@ -29,7 +29,7 @@ use tracing::{debug, instrument, warn};
 use super::ObligationCtxt;
 use crate::error_reporting::traits::suggest_new_overflow_limit;
 use crate::infer::InferOk;
-use crate::solve::inspect::{InspectGoal, ProofTreeInferCtxtExt, ProofTreeVisitor};
+use crate::solve::inspect::{InspectGoal, InferCtxtProofTreeExt, ProofTreeVisitor};
 use crate::solve::{SolverDelegate, deeply_normalize_for_diagnostics, inspect};
 use crate::traits::query::evaluate_obligation::InferCtxtExt;
 use crate::traits::select::IntercrateAmbiguityCause;

--- a/tests/ui/dst/dst-object-from-unsized-type.rs
+++ b/tests/ui/dst/dst-object-from-unsized-type.rs
@@ -1,4 +1,7 @@
 // Test that we cannot create objects from unsized types.
+//@ revisions: current next
+//@ ignore-compare-mode-next-solver (explicit revisions)
+//@[next] compile-flags: -Znext-solver
 
 trait Foo { fn foo(&self) {} }
 impl Foo for str {}

--- a/tests/ui/traits/next-solver/unsize-goal-mismatch-2.rs
+++ b/tests/ui/traits/next-solver/unsize-goal-mismatch-2.rs
@@ -1,0 +1,18 @@
+//@ check-pass
+//@ revisions: current next
+//@ ignore-compare-mode-next-solver (explicit revisions)
+//@[next] compile-flags: -Znext-solver
+// Test from trait-system-refactor-initiative#241:
+// Used to ICE in mir typeck because of ambiguity in the new solver.
+// The wrong (first) trait bound was selected.
+// This is fixed with new logic for unsizing coercions
+// that's independent from that of the old solver, which this test verifies.
+
+trait Super<T> {}
+trait Trait<T>: Super<T> + for<'hr> Super<&'hr ()> {}
+
+fn foo<'a>(x: Box<dyn Trait<&'a ()>>) -> Box<dyn Super<&'a ()>> {
+    x
+}
+
+fn main() {}

--- a/tests/ui/traits/next-solver/unsize-goal-mismatch.rs
+++ b/tests/ui/traits/next-solver/unsize-goal-mismatch.rs
@@ -1,0 +1,19 @@
+//@ check-pass
+//@ revisions: current next
+//@ ignore-compare-mode-next-solver (explicit revisions)
+//@[next] compile-flags: -Znext-solver
+// Test from trait-system-refactor-initiative#241:
+// Used to ICE in mir typeck because of ambiguity in the new solver.
+// The wrong (first) trait bound was selected.
+// This is fixed with new logic for unsizing coercions
+// that's independent from that of the old solver, which this test verifies.
+
+trait Super<'a> {}
+trait Trait<'a>: Super<'a> + for<'hr> Super<'hr> {}
+
+fn foo<'a>(x: Box<dyn Trait<'a>>) -> Box<dyn Super<'a>> {
+    x
+    //~^ ERROR type annotations needed: cannot satisfy `dyn Trait<'_>: Unsize<dyn Super<'_>>
+}
+
+fn main() {}

--- a/tests/ui/traits/next-solver/unsize-overflow.rs
+++ b/tests/ui/traits/next-solver/unsize-overflow.rs
@@ -1,0 +1,7 @@
+//@ compile-flags: -Znext-solver
+#![recursion_limit = "8"]
+
+fn main() {
+    let _: Box<dyn Send> = Box::new(&&&&&&&1);
+    //~^ ERROR overflow evaluating the requirement `Box<&&&&&&&i32>: CoerceUnsized<Box<dyn Send>>
+}

